src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -759,9 +759,32 @@ public override ImmutableArray<TypeParameterSymbol> TypeParameters
 
         #region Attributes
 
-        internal ImmutableArray<SyntaxList<AttributeListSyntax>> GetAttributeDeclarations()
+        /// <summary>
+        /// Gets all the attribute lists for this named type.  If <paramref name="quickAttributes"/> is provided
+        /// the attribute lists will only be returned if there is reasonable belief that 
+        /// the type has one of the attributes specified by <paramref name="quickAttributes"/> on it.
+        /// This can avoid going back to syntax if we know the type definitely doesn't have an attribute
+        /// on it that could be the one specified by <paramref name="quickAttributes"/>. Pass <see langword="null"/>
+        /// to get all attribute declarations.
+        /// </summary>
+        internal ImmutableArray<SyntaxList<AttributeListSyntax>> GetAttributeDeclarations(QuickAttributes? quickAttributes = null)
         {
-            return declaration.GetAttributeDeclarations();
+            // if the compilation has any global aliases to these quick attributes, then we have to return
+            // all the attributes on the decl.  For example, if there is a `global using X = Y;` and 
+            // then we have to return any attributes on the type as they might say `[Y]`.
+            if (quickAttributes != null)
+            {
+                foreach (var decl in this.DeclaringCompilation.MergedRootDeclaration.Declarations)
+                {
+                    if (decl is RootSingleNamespaceDeclaration rootNamespaceDecl)
+                    {
+                        if ((rootNamespaceDecl.GlobalAliasedQuickAttributes & quickAttributes) != 0)
+                            return declaration.GetAttributeDeclarations(quickAttributes: null);
+                    }
+                }
+            }
```

Comment:
Is it possible to verify that the performance issue is indeed resolved by this change (ie. there are no other performance bottlenecks on this common path)?

Reply:
i verified that we weren't going back to source.  but it's unclear if the entire bottleneck will be fixed.  we'll have to collect traces from the wild for that to know for sure.

Reply:
Glad to hear. Thanks!